### ISSUE TITLE
page server returned error: Failed to get reconstruct data

### ISSUE THREAD

**Issue by lubennikovaav** [neondatabase/neon#1215]
Weird error in pageserver. Maybe a duplicate
@LizardWizzard , FYI

```
main=> CREATE EXTENSION dblink;
ERROR:  could not read block 0 in rel 1663/16385/3467.0 from page server at lsn 0/0169BD38
DETAIL:  page server returned error: Failed to get reconstruct data 1663/16385/3467.0 "inmem-rel_1663_16385_3467_0_0_000000000169B638_FFFFFFFFFFFFFFFF" 0 0/169C418 None
```

The second call of the same command is ok.

**Comment by LizardWizzard** [comment]
I've observed the same error but with different actions. I've run pgbench init with scale 500 and everything was fine, but then I executed VACUUM FULL which caused weird behaviour. I see the following error in compute logs:
```
compute-node-proud-paper-996325		;
compute-node-proud-paper-996325	2022-02-04 23:13:42.862 GMT [35] LOG:  last-written LSN 7/D8136CA8 is ahead of last flushed LSN 7/D8134CD0
compute-node-proud-paper-996325	2022-02-04 23:13:42.862 GMT [35] STATEMENT:  VACUUM FULL
compute-node-proud-paper-996325		;
compute-node-proud-paper-996325	2022-02-04 23:13:42.873 GMT [35] ERROR:  could not read block 267520 in rel 1663/16385/49152.0 from page server at lsn 7/D8136CA8
compute-node-proud-paper-996325	2022-02-04 23:13:42.873 GMT [35] DETAIL:  page server returned error: Failed to get reconstruct data 1663/16385/49152.209 "rel_1663_16385_49152_0_209_00000005016EE5E8" 0 7/D8136CA8 None
compute-node-proud-paper-996325	2022-02-04 23:13:42.873 GMT [35] STATEMENT:  VACUUM FULL
compute-node-proud-paper-996325		;
4x			
compute-node-proud-paper-996325	2022-02-04 23:13:43.492 UTC [tokio-runtime-worker] INFO: serving /last_active GET request
compute-node-proud-paper-996325	2022-02-04 23:14:02.471 GMT [49] LOG:  [ZENITH_SMGR] libpqpagestore: connected to 'postgresql://(null):@zenith-1-ps-1.local:6400'
compute-node-proud-paper-996325	2022-02-04 23:14:03.494 UTC [tokio-runtime-worker] INFO: serving /last_active GET request
compute-node-proud-paper-996325	2022-02-04 23:14:06.235 GMT [50] LOG:  [ZENITH_SMGR] libpqpagestore: connected to 'postgresql://(null):@zenith-1-ps-1.local:6400' at character 31
compute-node-proud-paper-996325	2022-02-04 23:14:06.461 GMT [50] ERROR:  could not read block 2 in rel 1663/16385/49244.0 from page server at lsn 7/D8136CA8
compute-node-proud-paper-996325	2022-02-04 23:14:06.461 GMT [50] DETAIL:  page server returned error: Failed to get reconstruct data 1663/16385/49244.0 "rel_1663_16385_49244_0_0_0000000756169970" 2 7/D8138CD0 None
compute-node-proud-paper-996325	2022-02-04 23:14:06.461 GMT [50] STATEMENT:  
```
And then there was no message to the client (no error, no connection breakage, nothing), and compute was shut down after 5 minutes. It may be related to cancellation issues in proxy, e g #1209

**Comment by LizardWizzard** [comment]
connection being stuck after vacuum full is reproduced even without this error from pageserver:

```
  |   | compute-node-proud-paper-996325	2022-02-05 16:25:59.940 GMT [38] LOG:  last-written LSN 9/A39D0590 is ahead of last flushed LSN 9/A39D0248
-- | -- | --
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:25:59.940 GMT [38] STATEMENT:  VACUUM FULL
5x |   |   | compute-node-proud-paper-996325	2022-02-05 16:26:03.490 UTC [tokio-runtime-worker] INFO: serving /last_active GET request
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:26:24.273 GMT [68] LOG:  [ZENITH_SMGR] libpqpagestore: connected to 'postgresql://(null):@zenith-1-ps-1.local:6400'
56x |   |   | compute-node-proud-paper-996325	2022-02-05 16:26:28.490 UTC [tokio-runtime-worker] INFO: serving /last_active GET request
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:31:03.642 GMT [23] LOG:  received fast shutdown request
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:31:03.642 GMT [23] LOG:  aborting any active transactions
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:31:03.643 GMT [37] FATAL:  terminating connection due to administrator command
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:31:03.643 GMT [23] LOG:  background worker "logical replication launcher" (PID 31) exited with exit code 1
  |   |   | compute-node-proud-paper-996325	2022-02-05 16:31:03.644 GMT [30] FATAL:  terminating background worker "WAL proposer" due to administrator command
```

**Comment by LizardWizzard** [comment]
Another case in our staging cluster that holds results of perf tests: https://app.circleci.com/pipelines/github/zenithdb/zenith/4545/workflows/8a3eff34-0fd3-4458-81e0-b81211c3fd7c/jobs/42457

**Comment by antons-antons** [comment]
pgbench/vacuum issue is likely related to hio.c RelationAddExtraBlocks.
I'm adding some instrumentation as part of https://github.com/zenithdb/zenith/issues/626 which helps nailing down other root causes of reading stale pages (or pages without images/deltas on page servers) 

ping me if you get a repro of this

**Comment by LizardWizzard** [comment]
There are actually two issues, one which looks like a connection handling problem on the proxy<->compute side, and a page reconstruction failure from pageserver. Let this issue be about the page reconstruction failure.

I managed to reproduce connection managenent problem even after proxy async rewrite landed. So I'll create separate issue for it

**Comment by LizardWizzard** [comment]
And this issue sometimes reproduces for cluster which we use to store performance test results. See https://app.circleci.com/pipelines/github/zenithdb/zenith/4589/workflows/7007640d-a03c-4df3-863c-21564773118d/jobs/42991

**Comment by stepashka** [comment]
> Should be revisited once #1270 lands

it landed!

**Comment by shanyp** [comment]
@problame  is this relevant ?
